### PR TITLE
[WIP] Reworking the logins code to work with the new a-s API

### PR DESCRIPTION
**Pull request by bendk**
Here's my attempt at updating the a-c code to work with the [upcoming logins changes in a-s](https://github.com/mozilla/application-services/tree/logins-per-field-encryption).  I think this one is good enough that it's ready for review, even though there's at least a couple TODOs left (the example browser and integrating with the fenix code).

The main conceptual change is that we now differentiate between
different types of login data:
  - `LoginEntry` is a login that was entered by the user.  It's not tied
    to any database record
  - `Login` is a login stored in the database.  It has all the LoginEntry
    data, plus guid, timePasswordChanged, etc.
  - `EncryptedLogin` is a Login with the username/password data encrypted

This raises a question: which class should we use when adapting the API?

I used `LoginEntry` for the save login functionality, because I think it
fits better conceptually.  We're saving the data the user just entered
in the form, not updating any particular record.  In fact
SaveLoginDialogFragment will change which record to update as the user
alters the username textbox.

I also used `LoginEntry` for the login picker, solely because the
selection get passed through the GeckoView autocomplete code, which
means it gets converted to/from a GeckoView `LoginEntry`.  There's no way
to cleanly do with a `Login` because it has too many fields.

One place where this gets weird is when we're listing login entries and
there's duplicated login data because of sync.  Then we're potentially
creating 2 list items with exactly the same data.  But this is just a
reflection of the current UI, which will indeed present 2 items that
look exactly the same for the user to pick between.

One place I used `Login` was for the autofill feature.  Here the
situation seemed reversed from the GeckoView one.
`AbstractAutofillConfirmActivity` needs to handle a round-trip between
the 3rd party app and it uses the GUID to do that. I don't
really understand this code though, please correct me if I'm doing
something wrong.

At some point it would be nice to use `EncryptedLogin` in places
where we don't actually need the decrypted data.  For example
showing the user that there's an autofill available.  This would
potentially allow us to delay requiring the encryption key.

Change details:

- Replaced the shouldUpdateOrCreateAsync /
  potentialDupesIgnoringUsername logic with something that should be a
  lot simpler.

- Removed `LoginValidationDelegate` and `GeckoLoginStorageDelegate`.
  The regular `LoginsStorage` API is easy enough to use that we don't
  need to wrap it.

- Updated `SaveLoginDialogFragment` and simplified `update()` method.

- Test updates:
  - Added functions to create fake logins to test with
  - Removed the unittests for service-sync-logins.  At this point all it's
    just wrapper code, so it didn't seem useful to test it.
  - Updated `SaveLoginDialogFragmentTest.kt` and added more test cases.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by bendk** [comment]
Just rebased to bring in changes from `main`.  I wanted to update the sample-browser and fenix to work with this, but I had build issues.  I don't think they currently build against `application-services/main`, so I gave up on trying to get them to build with this branch.  Please correct me if I'm wrong on this one and I can try to get those to work as well.

**Comment by bendk** [comment]
Any updates on this one?  It would be great to know if I'm on the right track or not.

**Comment by grigoryk** [comment]
@bendk sorry for the delay, I'll take a look at this today/tomorrow!

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt [hunk 1]
```diff
@@ -73,23 +74,61 @@ internal class SaveLoginDialogFragment : PromptDialogFragment() {
...
     internal var password by SafeArgString(KEY_LOGIN_PASSWORD)
-
     private var validateStateUpdate: Job? = null
+    // List of logins already stored for this site.  We use this to determine
+    // how to save the new [LoginEntry] (add a new record, or update an existing record)
+    private var loginsForOrigin: Deferred<List<Login>>? = null
+
+    // Start the process to fetch `loginsForOrigin` if we need to.
+    // This makes sure that:
+    //   - We don't start the process before `feature.loginsStorage` is set
```
Comment (grigoryk) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt:94]: nit: maybe more specifically, "before `feature` is linked to this fragment". E.g. see `PromptFeature#reattachFragment`.

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt [hunk 2]
```diff
@@ -73,23 +74,61 @@ internal class SaveLoginDialogFragment : PromptDialogFragment() {
...
+    //   - We only run the process once
+    private fun fetchLoginsForOriginIfNeeded() {
+        val loginsStorage = feature?.loginsStorage
+        if (loginsStorage != null && loginsForOrigin == null) {
```
Comment (grigoryk) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt:98]: `this.loginsForStorage`, right? You're shadowing a class var just below, so make sure to be consistent with usage of `this`.

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt [hunk 3]
```diff
@@ -73,23 +74,61 @@ internal class SaveLoginDialogFragment : PromptDialogFragment() {
...
+            val loginsForOrigin = CoroutineScope(dispatcher).async {
+                loginsStorage.decryptLogins(loginsStorage.getByBaseDomain(origin))
+            }
+            loginsForOrigin.invokeOnCompletion {
+                // Make sure to run update() now that the logins have been fetched
+                CoroutineScope(Main).launch {
+                    update()
```
Comment (grigoryk) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt:105]: My first instinct is to move this elsewhere and restructure this code a bit. I.e. separate stuff that updates UI from stuff that manages and mutates local state.

As-is, you now have a side-effect to this function that changes UI state, which is a potential footgun. E.g. it's not clear now at what points UI will actually change - simply calling `findLogin` method can end up invoking this, which is _maybe_ what you want, but seems a bit surprising to me.

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt [hunk 4]
```diff
@@ -73,23 +74,61 @@ internal class SaveLoginDialogFragment : PromptDialogFragment() {
...
     private var validateStateUpdate: Job? = null
+    // List of logins already stored for this site.  We use this to determine
+    // how to save the new [LoginEntry] (add a new record, or update an existing record)
+    private var loginsForOrigin: Deferred<List<Login>>? = null
+
+    // Start the process to fetch `loginsForOrigin` if we need to.
+    // This makes sure that:
+    //   - We don't start the process before `feature.loginsStorage` is set
+    //   - We only run the process once
+    private fun fetchLoginsForOriginIfNeeded() {
```
Comment (grigoryk) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt:96]: `suspend`?

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt [hunk 5]
```diff
@@ -73,23 +74,61 @@ internal class SaveLoginDialogFragment : PromptDialogFragment() {
...
+                // Make sure to run update() now that the logins have been fetched
+                CoroutineScope(Main).launch {
+                    update()
+                }
+            }
+            this.loginsForOrigin = loginsForOrigin
+        }
+    }
+
+    private fun findLoginToUpdate(): Login? {
```
Comment (grigoryk) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt:112]: also, `suspend`..?

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt [hunk 6]
```diff
@@ -289,77 +333,38 @@ internal class SaveLoginDialogFragment : PromptDialogFragment() {
...
-                )
-            }
-            // Passing a copy of this potential dupes list, not a reference, so we know for certain
-            // that what shouldUpdateOrCreateAsync is using can't change.
-            validateDeferred = validationDelegate.shouldUpdateOrCreateAsync(
-                login,
-                potentialDupesList?.toList()
+    private fun update() {
+        // Figure out if we're going to be adding a login or updating an existing one
+        val loginToUpdate = findLoginToUpdate()
```
Comment (grigoryk) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt:338]: `findLoginToUpdate` can end up invoking `update` via `fetchLoginsForOriginIfNeeded`. You're one slip-up  away from a non-obvious infinite loop :)

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt [hunk 7]
```diff
@@ -289,77 +333,38 @@ internal class SaveLoginDialogFragment : PromptDialogFragment() {
...
+
+        if (loginToUpdate == null) {
+            // Adding a login (or we don't know what's happening because we're
+            // waiting for `loginsStorage.getByBaseDomain()`)
```
Comment (grigoryk) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt:342]: I don't think I follow. You're not calling getByBaseDomain in this class (unless i missed it). At this point, I would not expect such async behaviour - e.g. you can await for any data you need before update the UI.

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt [hunk 8]
```diff
@@ -289,77 +333,38 @@ internal class SaveLoginDialogFragment : PromptDialogFragment() {
...
-                        login
-                    ).await()
-                )
-            }
-            // Passing a copy of this potential dupes list, not a reference, so we know for certain
-            // that what shouldUpdateOrCreateAsync is using can't change.
-            validateDeferred = validationDelegate.shouldUpdateOrCreateAsync(
-                login,
-                potentialDupesList?.toList()
+    private fun update() {
```
Comment (grigoryk) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt:336]: This function should take `Login?` instead of invoking storage stuff. I think that's a good start to start cleaning this code up a bit.

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt [hunk 9]
```diff
@@ -289,77 +333,38 @@ internal class SaveLoginDialogFragment : PromptDialogFragment() {
...
+                negativeText = context?.getString(R.string.mozac_feature_prompt_never_save),
+                confirmText = context?.getString(R.string.mozac_feature_prompt_save_confirmation)
+            )
+        } else {
+            // Updating an existing login
+            setViewState(
+                headline = if (loginToUpdate.username == username) {
+                    // Normal case
+                    context?.getString(R.string.mozac_feature_prompt_login_update_headline)
+                } else {
```
Comment (grigoryk) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt:354]: This is a different check from before (we were doing isEmpty, not comparing usernames). ISTM that you're assuming that the only way for `loginToUpdate.username != username` is to have `loginToUpdate.username.isEmpty`. What guarantees this assumption? Is it the behaviour of `findLoginToUpdate`? If so, can you document that in a comment? Or simply switch this if statement to what it was before.

E.g. if this assumption doesn't hold you'll be displaying `add_username_headline` when we're actually changing a username on the record. It _is_ possible to change the username for an existing record, right?

**Review by grigoryk — CHANGES_REQUESTED** [review]
For sake of future PRs, I'll complain a bit to start :) There is a lot going on here, which makes it hard to follow all the different strands and actually review well.

An alternative PR would have been much smaller and just limited to just the encrypted/plaintext data switch - e.g. establishing when we deal with plaintext records, when we have encrypted records, and at which points the transformation happens. Then, we could have introduced further changes in follow-ups - moving validation logic into A-S, changing how storage is managed, etc.

You could have also structured this PR as multiple commits, where each concerns itself with a single task. Then review could be done commit-by-commit, making things much easier to follow with higher chances of success.

But alas, we have this PR now, so let's try this :) But please keep in mind that I'm basically certain I've missed something important.

Other than various code-related comments, I have two high-level questions:

1) Are you confident in the scope of this PR, and that it's all correct? We can still limit it to just the encrypted/decrypted changes, and keep all of the validation stuff in place, addressing it separately from the per-field encryption. Those are very different tasks with different goals and reasons for being.

2)
Why is this API structured differently from what we have in `autofill` for credit cards? Mainly, the API boundary there is much simpler - it's _always_ dealing with encrypted records; `autofill` lib provides a pair of functions to perform encryption/decryption, and provides key generation. And then it's responsibility of consumers to encrypt/decrypt stuff when necessary, and to manage the generated key.

That to me seems nice. A-S owns crypto bits, while its consumers own key management. The API surface is easy to reason about and predictable - it's _always_ ciphertext.

Having different approaches in different modules warrants an explanation. Is this accidental, or is there a good reason to do things differently here? If we have the same approach everywhere, we have a simpler system - things look and behave the same way when looking at different parts of the system, there's less complexity and less cognitive load.

I'd much prefer we stick to the same approach (and I prefer the one in `autofill`, seems like it worked out quite well!).

File: components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/SyncableLoginsStorage.kt [hunk 10]
```diff
@@ -166,129 +154,98 @@ class SyncableLoginsStorage(
...
+     * @throws [CryptoException] invalid encryption key
      */
-    @Throws(LoginsStorageException::class)
-    override suspend fun getPotentialDupesIgnoringUsername(login: Login): List<Login> =
-        withContext(coroutineContext) {
-            conn.getStorage().potentialDupesIgnoringUsername(login.toDatabaseLogin())
-                .map { it.toLogin() }
+    override suspend fun decryptLogins(logins: List<EncryptedLogin>): List<Login> = withContext(coroutineContext) {
+        logins.map {
+            mozilla.appservices.logins.decryptLogin(it.toAS(), key).toAC()
```
[the remark this answers is not in the document]
Comment (grigoryk) [reply components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/SyncableLoginsStorage.kt:243]: See my PR-level comment about the overall shape of the crypto stuff.

**Comment by bendk** [comment]
Thanks for the review!  I agree with your main point, which is that there's just too many changes here.  Let's discuss some ways it could be scaled back, and I'll try to explain some of the choices along the way.

First off, I think the encryption change doesn't really need to be exposed yet.  We can just decrypt the logins before returning them from the API methods.  There isn't any current code that uses encryted logins and it's not clear to me that the API is the right shape for the future world where we delay requesting the key.  Maybe instead of dealing with lists of encrypted logins, we would just want a `hasAnyLoginForBaseDomain()` method.

The second issue is the validation logic.  This is the one change I really want to push for. IMO, it's very hard to reason about the current logic, which makes any changes harder to do.  There's also something that feels bad about having to re-implement `potentialDupesExcludingUsername()`, which seems it shouldn't be in the API (OTOH it's not really that much work).

Related to the validation is the issue of `Login` vs `LoginEntry`.  I'd really love to get some good names here, because there's a subtle but important difference between those 2 things.  `Login` vs `LoginEntry` is clearly not communicating that to you.  I don't want to just swap the names, since we've gotten somewhat used to the current ones.  Do any of these pairs work for you: `LoginRecord` vs `Login`, `Login` vs `AutoFillLogin`, `Login` vs `LoginInfo`?

Why is the distinction important?  Because there's some confusing parts about how `Login` gets passed to `SaveLoginDialogFragment` and `LoginValidationDelegate`.  This login can have a `guid` field set, which to me should indicate that we're updating the login with that `guid` as the primary key.  However, depending on how the dupe-checking turns out, we might end up creating a new login or updating a different login.  When I try to picture `SaveLoginDialogFragment` as saving a `Login`, then the picture gets very murky.  It's clearer to think of it as saving the form entry.  For example, the update path in the current code goes: save a login -> match it to a duplicate -> merge the saved login into the dupe -> save the dupe.   In the new code this is: save a login entry -> find an existing login to update -> call `update(loginToUpdate.guid, loginEntry)` (not quite -- it actually calls addOrUpdate() in the end because I was worried about some corner cases but I still think the overall story is clearer).  In the current code, `guid` doesn't mean "primary key" of the login data like it normally does.  It's more like a tiebreaker in case of dupes.  In fact, I think we don't need `guid` at all and I ended up not using it in the login saving code, but I would love to go through some use cases with you to make sure.

Once I made that change it ended up causing a lot more though.  I thought if `SaveLoginDialogFragment` operates on a `LoginEntry`, then `SaveLoginPrompt` should also, which means that `GeckoAutocompleteStorageDelegate` too, and so on.  It ended up cascading all the way to `BasicLoginAdapter`.  Now that I think about it again, I realize I didn't need to do that.  I can just keep the others the same and have `SaveLoginDialogFragement` do a conversion.  I think that it might be useful to update that code later on, but it doesn't need to be part of this issue.

The other main changes were `LoginStorageDelegate`, `LoginValidationDelegate`, and the UI code inside `SaveLoginDialogFragment`.  I think there was good reasons to make those changes and I'll comment inline about each.  But if you're thinking it should be done in a different PR, or if it's not a good idea at all, I can back them out of this one.

Last thing: sorry about the big commit.  I was tunneling on getting the tests to pass before committing.  It should be split up more though, maybe I'll just do one component per commit and not worry about every test passing.

File: components/concept/storage/src/main/java/mozilla/components/concept/storage/LoginsStorage.kt [hunk 11]
```diff
@@ -141,155 +32,225 @@ data class Login(
...
-    val formActionOrigin: String? = null,
+    val formActionOrigin: String?,
     /**
      * The HTTP realm this login entry was requested for.
      * This only applies to non-form-based login entries.
      * It's derived from the WWW-Authenticate header set in a HTTP 401
      * response, see RFC2617 for details.
      */
-    val httpRealm: String? = null,
+    val httpRealm: String?,
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/concept/storage/src/main/java/mozilla/components/concept/storage/LoginsStorage.kt:42]: My reasoning here was that the A-C code doesn't need to construct a `Login` anymore, it only uses the ones that the A-S code returns.  The only exception was the tests, which I made the fixture functions for.  But those functions do feel a bit out of place in `support.test`.  I'll back these changes out.

File: components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/SyncableLoginsStorage.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (bendk) [reply components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/SyncableLoginsStorage.kt:243]: In general, the logins code is more complicated and this made us rethink how encryption should work.  I think the long-term goal might be to move credit-cards to being more like logins.

A good example is the de-duping code.  That needs to match the login data that's being saved against a list of existing logins.  It's nicer for that to be a list of decrypted logins rather than having a list of encrypted logins and decrypting the usernames for each pass.

Another one is autofill.  You need decrypt the username to display the list, then once the user selects an item you need to decrypt again.  It seemed to us that it was better do do a single  `List<EncryptedLogin> -> List<Login>` conversion, than to decrypt multiple times, or to have a `Login` class that might or might not have decrypted data

You bring up a really good point about the data marshalling though.  We should change the logins API so that only the username/password is sent across the FFI.  The code in a-s is setup to do exactly that.  I think the `concept.storage.LoginsStorage` interface could stay the same though.

File: components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/SyncableLoginsStorage.kt [hunk 12]
```diff
@@ -99,27 +88,26 @@ typealias InvalidKeyException = mozilla.appservices.logins.LoginsStorageExceptio
...
 ) : LoginsStorage, SyncableStore, AutoCloseable {
-    private val logger = Logger("SyncableLoginsStorage")
+    private val logger = Logger("LoginStorage")
     private val coroutineContext by lazy { Dispatchers.IO }
-
-    private val conn by lazy {
-        LoginStorageConnection.init(key = key, dbPath = context.getDatabasePath(DB_NAME).absolutePath)
-        LoginStorageConnection
+    private val store by lazy {
+        mozilla.appservices.logins.LoginStore(context.getDatabasePath(DB_NAME).absolutePath)
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/SyncableLoginsStorage.kt:103]: Yeah, I don't know why I made this change.  I'll revert it.

File: components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/Logins.kt [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.sync.logins
+
+import mozilla.components.concept.storage.*
+
+/// Convert between application-services data classes and the ones in
+/// concepts.storage.
+
+fun mozilla.appservices.logins.Login.toAC() = Login(
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/Logins.kt:12]: I really wanted to call them `into`, but there's multiple conversion functions for each class (`Login -> LoginEntry` and `Login -> GeckoView.Login`).  Can `into` be polymorphic based on the return value?  If not, I can use `toLogin`.  Although I thought that might be confusing from the other side, "why does Login have a toLogin method?"

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/autofill/GeckoAutocompleteStorageDelegate.kt [hunk 14]
```diff
@@ -24,12 +24,12 @@ import org.mozilla.geckoview.GeckoResult
...
  * Provides methods for retrieving [CreditCard]s from the underlying storage.
- * @param loginStorageDelegate An instance of [LoginStorageDelegate].
- * Provides read/write methods for the [Login] storage.
+ * @param loginsStorage An instance of [LoginsStorage].
+ * Provides read/write methods for the login storage.
  */
 class GeckoAutocompleteStorageDelegate(
     private val creditCardsAddressesStorageDelegate: CreditCardsAddressesStorageDelegate,
-    private val loginStorageDelegate: LoginStorageDelegate
+    private val loginsStorage: LoginsStorage
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/autofill/GeckoAutocompleteStorageDelegate.kt:32]: But doesn't `SyncableLoginsStorage` also have a lazy reference to the store/connection?  I didn't understand why we needed 2 layers of laziness there.  I could definitely just be missing something here though.

File: components/service/sync-logins/README.md [hunk 15]
```diff
@@ -43,136 +43,33 @@ implementation "org.mozilla.components:fxa:{latest-version}
 
 * Android 6.0 is temporary not supported and will probably crash the application.
 
-### Example
+## API
 
-How to display the data before syncing, sync, and then display the data after
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/service/sync-logins/README.md:48]: It seemed like they were getting copied too much.  I had already copied the doc strings from the a-s repo to `concept/storage/src/main/java/mozilla/components/concept/storage/LoginsStorage.kt` and didn't want to copy them again to the README.  Duplicate documentation makes me nervous that they're going to get out of sync and then it's not clear which one is the correct one.  I'd rather have a breadtrail back to the canonical version.

File: components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegate.kt [hunk 16]
```diff
@@ -1,128 +0,0 @@
...
-    ): Deferred<Result> {
-        return scope.async {
-            val potentialDupesList =
-                potentialDupes ?: storage.value.getPotentialDupesIgnoringUsername(newLogin)
-            val foundLogin = if (potentialDupesList.isEmpty()) {
-                // If list is empty, no records match -> create
-                null
-            } else {
-                // Matching guid, non-blank matching username -> update
-                potentialDupesList.find { it.guid == newLogin.guid && it.username == newLogin.username }
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegate.kt:41]: Yes, now both the `add()` and `update()` functions do fixups/validation.  https://github.com/mozilla/application-services/blob/logins-per-field-encryption/components/logins/src/db.rs#L510 and also the `find_login_to_update()` function simplifies the dupe-checking code.  I'll make sure to add a comments for this.

File: samples/browser/src/main/java/org/mozilla/samples/browser/storage/DummyLoginsStorage.kt [hunk 17]
```diff
@@ -4,71 +4,156 @@
...
+    timeCreated = timeCreated,
+    timePasswordChanged = timePasswordChanged,
+    secureFields = encryptLoginData(entry.username, entry.password),
+)
+
 /**
  * A dummy [LoginsStorage] that returns a fixed set of fake logins. Currently sample browser does
  * not use an actual login storage backend.
  */
 class DummyLoginsStorage : LoginsStorage {
```
[the remark this answers is not in the document]
Comment (bendk) [reply samples/browser/src/main/java/org/mozilla/samples/browser/storage/DummyLoginsStorage.kt:76]: Agreed, it's also an extra thing that needs to change when the logins APi changes.  If it's easy to just replace the dummy storage layer, maybe I'll try to do that.  Is it okay for  the sample browser to depend on a-s?

File: components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/Logins.kt [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+
+fun LoginEntry.toAS() = mozilla.appservices.logins.LoginEntry(
+    fields = mozilla.appservices.logins.LoginFields(
+        origin = origin,
+        formActionOrigin = formActionOrigin,
+        httpRealm = httpRealm,
+        usernameField = usernameField,
+        passwordField = passwordField,
+    ),
+    secFields = mozilla.appservices.logins.SecureLoginFields(
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/Logins.kt:49]: This was another name that was hard to pick.  "secFields" are the fields that are secured when at rest as opposed to "fields".  The idea is to group the fields together.  It's the login type that determines of they're encrypted or not (`EncryptedLogin` vs `Login` and `LoginEntry`).

File: components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/Logins.kt [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+
+package mozilla.components.service.sync.logins
+
+import mozilla.components.concept.storage.*
+
+/// Convert between application-services data classes and the ones in
+/// concepts.storage.
+
+fun mozilla.appservices.logins.Login.toAC() = Login(
+    guid = record.id,
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/Logins.kt:13]: I this was related to FFI issues.  `id` means the string, while `guid` means a `Guid` struct.  The struct used in the Rust `Login` struct, but we need to pass a string over the FFI.  I'm not sure I love the naming, but maybe it's okay for now.  I think the plan is to update UniFFI to support wrapper types so we could just use `Guid` directly in the UDL.

File: components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/Logins.kt [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+    origin = fields.origin,
+    formActionOrigin = fields.formActionOrigin,
+    httpRealm = fields.httpRealm,
+    usernameField = fields.usernameField,
+    passwordField = fields.passwordField,
+    timesUsed = record.timesUsed,
+    timeCreated = record.timeCreated,
+    timeLastUsed = record.timeLastUsed,
+    timePasswordChanged = record.timePasswordChanged,
+    secureFields = secFields,
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/Logins.kt:38]: I think part of it was the extra decryption step and another part was how to manage the data.  If you decrypt the username, how do you store the decrypted username with the login so you don't need to keep decrypting it?  I think the alternative would be to have a `Login` type where each piece of data was sometimes encrypted and sometimes decrypted.  But that can get confusing quickly.  If you have a list of logins, how do you know which data is decrypted or not?

It seemed simpler to do all the decryption in one pass and the goal would be that that pass would happen when the user clicks on "suggested logins" and  we needed to display the login list to show the user.  This is also why `decryptLogins` works on a list rather than an individual login (of course if you wanted a single-item function, we could definitely add that).

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt (hunk 3 above)
Comment (bendk) [reply components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt:105]: I like that idea a lot and it would also make the testing easier.  I was afraid to do that when I was coding that, mostly because I'm not sure how the Android lifecycle works.  If I split out a class that handles the dupe-checking, when would I construct it? It would need to be after the feature is set but before the view needs updating.  Would `onViewCreated` work?  Is that called when Android is recreating the fragment?

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt (hunk 6 above)
Comment (bendk) [reply components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt:338]: This was from the fear I mentioned earlier about not being sure how to reliably start up the `loginsForOrigin` deferred, so I added this as a just in case `onCreateView()` hadn't already started it up.  I'll make sure to clean this up for the next pass.

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt (hunk 7 above)
Comment (bendk) [reply components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt:342]: It's called at the top to setup `loginsForOrigin`.  For me, it seemed slightly simpler to have `update()` not be suspend -- it updates the UI immediately using the currently available data -- and then make sure to call `update()` once the deferred completes.  That way you don't need to cancel the previous job at the top and maybe it simplifies storing the potential dupes list.  That's a very minor preference though.

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt (hunk 9 above)
Comment (bendk) [reply components/feature/prompts/src/main/java/mozilla/components/feature/prompts/dialog/SaveLoginDialogFragment.kt:354]: I believe the only way that this code will change a username is when it fills it in for a login with a blank username (both currently and after the suggested changes).  `findLoginToUpdate` first tries to find a login with a matching username to update.  If it can't find that it tries to find a login with an empty username to fill in.  If neither of those are true, then it returns None and we should add a new login.  I'll add some documentation around this.

Couple of notes here:
  - In the case where the user is saving an empty username and there's an existing login with an empty username, the behavior has changed.  Before it would say use the add username headline, now it uses the update login headline.  I'm pretty sure that's just a bug fix and I checked it with Elise, but tell me if that's not correct.
  - Before the GUID sometimes modified this behavior now it doesn't.

File: components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/Logins.kt (hunk 20 above)
[the remark this answers is not in the document]
Comment (bendk) [reply components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/Logins.kt:38]: BTW, I also would personally push for storing the username plaintext.  I don't really feel like it's sensitive data in general.  But I think that would require a larger discussion.

**Comment by bendk** [comment]
Just opened #10987, which I hope should be easier to review.  There's definitely many less lines changed.  I'm still wondering if the delegate / `SaveLoginDialogFragment` code could be updated, but that can be a different discussion.  I think this one can be closed now.
